crates/cairo-addons/src/vm/hint_definitions/curve.rs

```diff
@@ -175,46 +288,89 @@ pub fn compute_y_from_x_hint() -> Hint {
+                    if has_same_parity {
+                        sqrt_felt
+                    } else {
+                        -sqrt_felt
+                    }
                 } else {
-                    -sqrt_felt
-                }
-            } else {
-                (rhs_felt * g_felt).sqrt().unwrap().0
-            };
-
-            let sqrt_biguint = element_to_biguint(&square_root);
-            Uint384::split(&sqrt_biguint).insert_from_var_name(
-                "y_try",
-                vm,
-                ids_data,
-                ap_tracking,
-            )?;
-            write_collection_from_var_name(
-                "is_on_curve",
-                &[is_on_curve.into(), Felt252::ZERO, Felt252::ZERO, Felt252::ZERO],
-                vm,
-                ids_data,
-                ap_tracking,
-            )?;
+                    (rhs_felt * g_felt).sqrt().unwrap().0
+                };
+
+                let sqrt_biguint = element_to_biguint(&square_root);
+                Uint384::split(&sqrt_biguint).insert_from_var_name(
+                    "y_try",
+                    vm,
+                    ids_data,
+                    ap_tracking,
+                )?;
+                write_collection_from_var_name(
+                    "is_on_curve",
+                    &[is_on_curve.into(), Felt252::ZERO, Felt252::ZERO, Felt252::ZERO],
+                    vm,
+                    ids_data,
+                    ap_tracking,
+                )?;
+            }
+            if is_bn254_p {
+                let rhs_felt =
+                    FieldElement::<BN254PrimeField>::from_hex_unchecked(&rhs.to_str_radix(16));
+                let g_felt =
+                    FieldElement::<BN254PrimeField>::from_hex_unchecked(&g.to_str_radix(16));
+
+                let is_on_curve = is_quad_residue(&rhs, &p);
+                let square_root = if is_on_curve {
+                    let sqrt_felt = rhs_felt.sqrt().unwrap().0;
+                    let has_same_parity = (v % 2_u32) == (element_to_biguint(&sqrt_felt) % 2_u32);
+                    if has_same_parity {
+                        sqrt_felt
+                    } else {
+                        -sqrt_felt
+                    }
+                } else {
+                    (rhs_felt * g_felt).sqrt().unwrap().0
+                };
+
+                let sqrt_biguint = element_to_biguint(&square_root);
+                Uint384::split(&sqrt_biguint).insert_from_var_name(
+                    "y_try",
+                    vm,
+                    ids_data,
+                    ap_tracking,
+                )?;
+                write_collection_from_var_name(
+                    "is_on_curve",
+                    &[is_on_curve.into(), Felt252::ZERO, Felt252::ZERO, Felt252::ZERO],
+                    vm,
+                    ids_data,
+                    ap_tracking,
+                )?;
+            }
             Ok(())
```

Comment:
You could do an early return in the if is_secp256k1_p and remove the second if (add a comment only as